pkg/serving/v1alpha1/client_test.go

```diff
@@ -0,0 +1,320 @@
+			notExistingServiceName = "no-service"
+		)
+		var (
+			service *v1alpha1.Service
+		)
+
+		BeforeEach(func() {
+			service = newService(serviceName)
+			serving.AddReactor("get", "services",
+				func(a client_testing.Action) (bool, runtime.Object, error) {
+					name := a.(client_testing.GetAction).GetName()
+					// Sanity check
+					Expect(name).ToNot(BeEmpty())
+					Expect(a.GetNamespace()).To(Equal(namespace))
+					if name == serviceName {
+						return true, service, nil
+					}
+					return true, nil, errors.NewNotFound(v1alpha1.Resource("service"), name)
+				})
+		})
+
+		Context("with a name that exists", func() {
+			It("returns a service with this name and without error", func() {
+				getService, err := knClient.GetService(serviceName)
+				Expect(err).To(BeNil())
+				Expect(getService.Name).To(Equal(service.Name))
+				validateGvk(getService)
+			})
+		})
+
+		Context("with a name that does not exist", func() {
+			It("returns a nil object and error with a descriptive error message", func() {
+				getService, err := knClient.GetService(notExistingServiceName)
+				Expect(getService).To(BeNil())
+				Expect(err).ToNot(BeNil())
+				Expect(err.Error()).To(ContainSubstring(notExistingServiceName))
+				Expect(err.Error()).To(ContainSubstring("not found"))
+			})
+		})
+	})
+
+	// -----------------------------------------------------------------------------------------
+	Describe("Listing services", func() {
+		var (
+			service1 = newService("service-1")
+			service2 = newService("service-2")
+		)
+
+		BeforeEach(func() {
+			serving.AddReactor("list", "services",
+				func(a client_testing.Action) (bool, runtime.Object, error) {
+					Expect(a.GetNamespace()).To(Equal(namespace))
+					return true, &v1alpha1.ServiceList{Items: []v1alpha1.Service{*service1, *service2}}, nil
+				})
+		})
+
+		It("returns a list of services", func() {
+			listServices, err := knClient.ListServices()
+			Expect(err).To(BeNil())
+			Expect(listServices.Items).To(HaveLen(2))
+			Expect(listServices.Items[0].Name).To(Equal("service-1"))
+			Expect(listServices.Items[1].Name).To(Equal("service-2"))
+			validateGvk(listServices)
+			validateGvk(&listServices.Items[0])
+			validateGvk(&listServices.Items[1])
+		})
+	})
+
+	// -----------------------------------------------------------------------------------------
+	Describe("Creating a service", func() {
+		var (
+			serviceNew = newService("new-service")
+		)
+
+		BeforeEach(func() {
+			serving.AddReactor("create", "services",
+				func(a client_testing.Action) (bool, runtime.Object, error) {
+					Expect(a.GetNamespace()).To(Equal(namespace))
+					name := a.(client_testing.CreateAction).GetObject().(metav1.Object).GetName()
+					if name == serviceNew.Name {
+						serviceNew.Generation = 2
+						return true, serviceNew, nil
+					}
+					return true, nil, fmt.Errorf("error while creating service %s", name)
+				})
+		})
+
+		Context("without error", func() {
```

Comment:
Nice use of context overall, that's precisely where Ginkgo shines. When you run the tests and there are failures makes it easy to understand what's expected and where the failures are occurring.